### PR TITLE
Implement bytecode callbacks with static, never-modified bytecode

### LINKED ISSUE #13512: GC crashes under bytecode
The [Gc torture test](https://github.com/ocaml-multicore/multicoretests/pull/469) has come to a point, where it can trigger a Gc crashing bug with the bytecode interpreter. This can trigger on both on `5.2.0`, on the `5.3` branch, and on `trunk`. I have not tried on 5.0.0 or 5.1.1 yet.

The test itself triggers list/string/array/bigarray allocations in both the parent domain and in two child domains and checks consistency of the results from `Gc.quick_stat`, etc. In between test iterations, we explicitly call `Gc.major` in single-domain mode in an attempt to reset the heap to a reasonable state. The test also uses a single call to a stub at start-up to support testing with `OCAMLRUNPARAM="s=4096"` set, but that can removed if unused.

This issue may or may not be the same bug observed in #13402.
As this test doesn't involve the `Obj`-using `Domain.DLS` I've gone for reporting it separately.

On GitHub actions CI I've observed these crashes with Ubuntu and glibc. For an example 5.3 crash see here:
https://github.com/ocaml-multicore/multicoretests/actions/runs/11074833760/job/30774502500?pr=469
Locally it seems easier to reproduce with musl C though.

To reproduce:
```
$ sudo apt-get install musl-tools

$ opam switch create . --empty
$ opam install . --inplace-build ocaml-option-musl ocaml-option-bytecode-only
$ opam install qcheck-core

$ git clone -b gc-test-musl-bytecode-repro https://github.com/ocaml-multicore/multicoretests.git
$ cd multicoretests
$ OCAMLRUNPARAM="o=20" dune exec src/gc/stm_tests_impl.exe --profile=debug-runtime -- -v
```

Above I suggest running with a reduced `space_overhead` and the debug runtime, but the bug can also trigger without these.

Here's a the result of a fresh run on my Linux box:
```
$ OCAMLRUNPARAM="o=20" dune exec src/gc/stm_tests_impl.exe --profile=debug-runtime -- -v
### OCaml runtime: debug mode ###   
### set OCAMLRUNPARAM=v=0 to silence this message
Page size: 4096
random seed: 57333975
generated error fail pass / total     time test name
[ ]  874    0    0  874 / 1000   500.8s STM implicit Gc test parallelBus error (core dumped)
```

with the following, not-so-descriptive stack trace:
```
(gdb) thread apply all bt

Thread 3 (LWP 1044096):
#0  0x00007fa68ff14b8c in memset () from /lib/ld-musl-x86_64.so.1
Backtrace stopped: Cannot access memory at address 0x7fa67cec89a8

Thread 2 (LWP 956593):
#0  0x00007fa68ff14bf8 in memset () from /lib/ld-musl-x86_64.so.1
#1  0x00007fa68ff05d35 in ?? () from /lib/ld-musl-x86_64.so.1
#2  0x00007fa67fc52b74 in ?? ()
#3  0x0000000000000000 in ?? ()

Thread 1 (LWP 956567):
#0  caml_interprete (prog=<optimized out>, prog_size=<optimized out>) at runtime/interp.c:829
#1  0x000060172030b6fd in caml_startup_code_exn (pooling=0, argv=0x7ffff57a4598, section_table_size=3833, section_table=0x60172032b020 <caml_sections> "\204\225\246\276", data_size=22927, data=0x60172032bf20 <caml_data> "\204\225\246\276", code_size=539264, code=0x6017203318c0 <caml_code>) at runtime/startup_byt.c:659
#2  caml_startup_code_exn (code=0x6017203318c0 <caml_code>, code_size=539264, data=0x60172032bf20 <caml_data> "\204\225\246\276", data_size=22927, section_table=0x60172032b020 <caml_sections> "\204\225\246\276", section_table_size=3833, pooling=0, argv=0x7ffff57a4598) at runtime/startup_byt.c:592
#3  0x000060172030b742 in caml_startup_code (code=code@entry=0x6017203318c0 <caml_code>, code_size=code_size@entry=539264, data=data@entry=0x60172032bf20 <caml_data> "\204\225\246\276", data_size=data_size@entry=22927, section_table=section_table@entry=0x60172032b020 <caml_sections> "\204\225\246\276", section_table_size=section_table_size@entry=3833, pooling=0, argv=0x7ffff57a4598) at runtime/startup_byt.c:673
#4  0x00006017202d5052 in main (argc=<optimized out>, argv=<optimized out>) at camlprim.c:26518
```

### PR DESCRIPTION
**Pull request by xavierleroy**
This is an alternative to #13549.  See this PR for the detailed description of the problem that needs to be fixed.

Like #13549, the present PR avoids modifying bytecode dynamically, and making it thread-local.

Unlike #13549, it preserves the ability to perform a callback on N arguments in one go, instead of having to cut it in slices of 1 to 3 arguments.

### CODE REVIEW

**Review by ghost — APPROVED** [review]
This also makes the failures disappear and looks good to me.

**Comment by stedolan** [comment]
This looks good! This approach does seem cleaner than either modifying bytecode or slicing up callbacks into 3-argument chunks, so I think this is a better fix than #13549.

However, I think there's a bug in the current version: the old code, going through `Instruct(APPLY)`, used to do a `goto check_stacks` before executing the code from the callback. I think this stack check is important, as I think it's possible to reach caml_callback without much available stack (there's no stack check before `Instruct(C_CALLn)`, so you can enter C code having already consumed part of your stack).

I think there's a simple fix: the interpreter should do a `goto check_stacks` right before entering the main loop, so that a stack check occurs before any instructions are executed.

**Comment by shindere** [comment]
Stephen Dolan (2024/10/16 02:36 -0700):
> I think there's a simple fix: the interpreter should do a `goto
> check_stacks` right before entering the main loop, so that a stack
> check occurs before any instructions are executed.

Won't that be too often then? I suppose many instructions do actually
not change the stack and it then feels a pity to me to do a check before
they are executed, leading to a performance loss. Or am I missing
something maybe?

**Comment by stedolan** [comment]
I'm not suggesting a check inside the loop, before every instruction - that would indeed cause performance loss. I'm suggesting a single check, _before_ the loop is entered, which would be once per invocation of `caml_bytecode_interpreter`.

**Comment by shindere** [comment]
Stephen Dolan (2024/10/16 05:55 -0700):
> I'm not suggesting a check inside the loop, before every instruction -
> that would indeed cause performance loss. I'm suggesting a single
> check, _before_ the loop is entered, which would be once per
> invocation of `caml_bytecode_interpreter`.

Ah okay, sorry I misunderstood. And you think such a check would be
enough, then?

**Comment by NickBarnes** [comment]
> I think this stack check is important, as I think it's possible to reach caml_callback without much available stack (there's no stack check before `Instruct(C_CALLn)`, so you can enter C code having already consumed part of your stack).

On this subject, shouldn't `caml_callbackN_exn` do a stack limit check before subtracting `narg + 4` from `sp`? Maybe I'm missing something obvious about the bytecode stacks.

**Comment by xavierleroy** [comment]
> the interpreter should do a goto check_stacks right before entering the main loop

Well spotted!  See the latest commit on this PR.

> shouldn't caml_callbackN_exn do a stack limit check before subtracting narg + 4 from sp

Right now we  rely on the "red zone" at the bottom of the stack to absorb the arguments.  There's an assertion but it's not up to date (it assumes 256 words of red zone while it's only 32 words nowadays).  Maybe it would be better to check and resize the stack in caml_callbackN_exn.  Will look into it in a couple of days.

**Comment by NickBarnes** [comment]
Ah, I see, `Stack_threshold_words` ?

**Comment by xavierleroy** [comment]
Yes, `Stack_threshold_words`.

I pushed an alternative to @stedolan's suggestion, based on @NickBarnes' remark, where `caml_callbackN` checks for stack space and reallocates the stack if needed.  I doubt the test suite exercises this code, though.  I'll try to come up with a test later this week.

**Comment by xavierleroy** [comment]
I added a test that triggers the stack resizing, and a Changes entry.  This PR is ready for a final review.

File: runtime/callback.c [hunk 1]
```diff
@@ -83,54 +83,54 @@ Caml_inline void restore_stack_parent(caml_domain_state* domain_state,
...
+  CAMLparam1(closure); /* no need to register args as roots, see below */
   CAMLlocal1(cont);
   value res;
   caml_domain_state* domain_state = Caml_state;
 
-  CAMLassert(narg + 4 <= 256);
-  domain_state->current_stack->sp -= narg + 4;
+  /* Ensure there's enough stack space */
+  intnat req = narg + 3 + Stack_threshold_words;
+  if (domain_state->current_stack->sp - req <
```
Comment (ghost) [inline runtime/callback.c:107]: This almost calls for an internal version of `caml_ensure_stack_capacity`, but that can be done later if worth doing.

**Comment by jmid** [comment]
After a good round of testing this afternoon I can confirm that this fixes both #13512 and #13402.
Thank you both!

**Comment by xavierleroy** [comment]
Cleaned up the history.  Rebased to fix conflict.  Knelt in front of check-typo.  Time to merge!

**Comment by Octachron** [comment]
Cherry-picked to the 5.2 branch as 71d85e068d in prevision of the upcoming 5.2.1 release.

**Pull request merged**